### PR TITLE
Update Loadbalancer when there is a change in endpoints of a service of type=LoadBalancer

### LINKED ISSUE #44203: UpdateLoadBalancer is not called for a service with type=LoadBalancer when pods are added/deleted
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

kubernetes external-loadbalancer
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

**Kubernetes version** (use `kubectl version`):

```
stack@kube-master:~$ sudo kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:57:05Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:52:01Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

```

**Environment**:
- **Cloud provider or hardware configuration**: Custom cloud provider
- **OS** (e.g. from /etc/os-release): 
```
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
```
- **Kernel** (e.g. `uname -a`): 

`Linux kube-master 4.4.0-57-generic #78-Ubuntu SMP Fri Dec 9 23:50:32 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

**What happened**:
**My setup:**
1 master node and 1 minion node

I am creating a new cloud provider plugin, to load-balance a service. The plugin implements LoadBalancer interface and is able to create a load balancer. This plugin adds pod IPs of the service as members to the load balancer. 

The problem is when the underlying replication controller of the service is scaled up/down , UpdateLoadBalancer function in the plugin is not invoked. After looking into code I found out that this function will only be called when a minion node is asssociated/disassociated with a service. 

**What you expected to happen**:
Whenever there is a change in the service (eg. service endpoints changed when replication controller is scaled up/down), UpdateLoadBalancer function should be invoked.

**How to reproduce it** (as minimally and precisely as possible):
In a 2 node setup, add a service and the scale up the pods to see this issue. 

**Anything else we need to know**:
I understand that this might not be an issue for other cloud providers as they use NodePort to load balance. There can be use cases where one wants to use the pod IPs to load balance.

### PR DESCRIPTION
**Pull request by spradeepv**
**What this PR does / why we need it**:
 This PR compares the endpoints of a old service object with a new service object, inside a watcher in servicecontroller.go. If there is a difference in the count, UpdateLoadBalancer() function of the provider's loadbalancer implementation is invoked for a service of type=LoadBalancer.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #44203

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by spradeepv** [comment]
Signed CLA

**Comment by xiangpengzhao** [comment]
@spradeepv Could you please amend your PR title and commit message to make them meaningful?

**Comment by MrHohn** [comment]
/release-note-none

@k8s-bot ok to test

cc @freehan @nicksardo

File: pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -125,16 +127,33 @@ func New(
...
 
 	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: s.enqueueService,
 			UpdateFunc: func(old, cur interface{}) {
+				endpointsChanged := false
 				oldSvc, ok1 := old.(*v1.Service)
 				curSvc, ok2 := cur.(*v1.Service)
+				oldEndpoints := s.endpoints[oldSvc.ObjectMeta.Name]
+				newEndpoints := s.getNumberOfEndpoints(oldSvc)
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller.go]: I don't think comparing number of endpoints is enough. What if one endpoint is removed while another is added?

**Review by MrHohn — COMMENTED** [review]
There is currently no requirement on any other cloud providers to call UpdateLoadBalancer when endpoints are updated. I could imagine in future we may want endpoints changes to trigger some sort of LB updates as well. Though this might be very much diverged in each cloud provider, I'm not sure if UpdateLoadBalancer() is the right entry for this functionality.

For now, this seems to introduce unnecessary function calls to all other cloud providers.

File: pkg/controller/service/servicecontroller.go (hunk 1 above)
Comment (sudhakso) [inline pkg/controller/service/servicecontroller.go]: We may not want to check endpoints when the service is being created first time. Services that gets updated less frequently should see no additional overhead.

File: pkg/controller/service/servicecontroller.go [hunk 2]
```diff
@@ -125,16 +127,33 @@ func New(
...
+		endpoints: make(map[string]int),
 	}
 
 	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: s.enqueueService,
 			UpdateFunc: func(old, cur interface{}) {
+				endpointsChanged := false
 				oldSvc, ok1 := old.(*v1.Service)
 				curSvc, ok2 := cur.(*v1.Service)
```
Comment (sudhakso) [inline pkg/controller/service/servicecontroller.go]: Can these checks be includes as part of 'needsUpdate' function? That way we consolidate the decision making through that interface.

File: pkg/controller/service/servicecontroller.go [hunk 3]
```diff
@@ -125,16 +127,33 @@ func New(
...
+				oldEndpoints := s.endpoints[oldSvc.ObjectMeta.Name]
+				newEndpoints := s.getNumberOfEndpoints(oldSvc)
+				if oldEndpoints != newEndpoints {
+					endpointsChanged = true
+				s.endpoints[oldSvc.ObjectMeta.Name] = newEndpoints
+				}
 				if ok1 && ok2 && s.needsUpdate(oldSvc, curSvc) {
 					s.enqueueService(cur)
+				} else if wantsLoadBalancer(curSvc) && endpointsChanged {
```
Comment (sudhakso) [inline pkg/controller/service/servicecontroller.go]: Can we move the calling part of UpdateLoadBalancer into createLoadBalancerIfNeeded(..) function? That would consolidate code that does action at one place.

File: pkg/controller/service/servicecontroller.go [hunk 4]
```diff
@@ -125,16 +127,33 @@ func New(
...
+					nodes, err := s.nodeLister.List(labels.Everything())
+					err = s.balancer.UpdateLoadBalancer(s.clusterName, curSvc, nodes)
+					if err == nil {
+						s.eventRecorder.Event(curSvc,
+							api.EventTypeNormal, 
+							"UpdatedLoadBalancer", 
+							"Updated load balancer with new Endpoints")
```
Comment (sudhakso) [inline pkg/controller/service/servicecontroller.go]: Can we handle event description in more generic way to handle 'add' and 'delete' members? This gives me an impression that it is considering addition of new endpoints only.

**Review by sudhakso — CHANGES_REQUESTED** [review]

File: pkg/controller/service/servicecontroller.go (hunk 1 above)
Comment (sudhakso) [inline pkg/controller/service/servicecontroller.go]: @MrHohn We need this implementation to handle updates for services of type=loadbalancer. And that's a choice that each Cloud Provider would have to make (that should be made configurable), if it uses 'nodeIPs' versus 'podIPs' as the LB members for such services.

I think it is reasonable to implement a check to decide whether the implementation uses NodeIp versus PodIp, and design the corse of this implementation.

File: pkg/controller/service/servicecontroller.go (hunk 2 above)
Comment (spradeepv) [inline pkg/controller/service/servicecontroller.go]: Done

File: pkg/controller/service/servicecontroller.go (hunk 1 above)
Comment (spradeepv) [inline pkg/controller/service/servicecontroller.go]: @MrHohn I have addressed your comment

File: pkg/controller/service/servicecontroller.go (hunk 3 above)
Comment (spradeepv) [inline pkg/controller/service/servicecontroller.go]: Done

File: pkg/controller/service/servicecontroller.go (hunk 4 above)
Comment (spradeepv) [inline pkg/controller/service/servicecontroller.go]: Done

**Comment by spradeepv** [comment]
Closing this pull request for now

**Comment by andelwu** [comment]
This is helpful.

**Pull request closed without merging**